app/controllers/application_controller.rb

```diff
@@ -254,12 +254,25 @@ def pull_pinned_links_and_hot_questions
         PinnedLink.where(active: true).where('shown_before IS NULL OR shown_before > NOW()').all
       end
     end
+
+    # We need to filter out pinned links that aren't posts
+    # If pinned_post_ids contains nil, then the query formed is
+    # SELECT WHERE NOT IN (... NULL)
+    # In SQL, when the rhs of IN contains NULL, it returns NULL if not present, rather than false
+    # Secondly, NOT NULL evaluates to NULL
+    # These two counterintuitive behaviors mean that if NULL is in pinned_post_ids,
+    # Post not pinned => NOT IN pinned_post_ids = NOT NULL = NULL => Not Selected
+    # Post pinned => NOT IN pinned_post_ids = NOT TRUE = FALSE => Not Selected
+    # I.e., if pinned_post_ids contains null, the selection will never return records
+    pinned_post_ids = @pinned_links.pluck(:post_id).select(&:present?)
```

Comment:
We should use `map` here instead of `pluck` - performance is usually better. I'd also (opinion alert) use `reject` rather than `select` - it narrates what you're trying to do a bit more clearly.

```suggestion
    pinned_post_ids = @pinned_links.map(&:post_id).reject(&:nil?)
```